math/arctangent.go

```diff
@@ -1,354 +1,268 @@
-	calculatingPrecision := precision + defaultExtraPrecision
-	x := decimal.WithPrecision(calculatingPrecision).Copy(xValue)
+	x0 := alias(z, x)
 
-	//since smaller values converge faster
+	// since smaller values converge faster
 	// we'll use argument reduction
-	//if |xValue| > 1 then set x = 1/xValue
+	// if |x| > 1 then set x = 1/x
 	segment := 0
-	if one.CmpAbs(xValue) < 0 {
+	if x.CmpAbs(one) > 0 {
 		segment = 2
-		x = x.Quo(one, x)
-	} else if decimal.New(577350, 6).CmpAbs(xValue) < 0 {
-		// if |xValue| > sqrt(3)/3  (we approx to 0.577350)
+		ctx.Quo(x0, one, x)
+	} else if x.CmpAbs(sqrt3_3) > 0 {
+		// if |x| > sqrt(3)/3 (approximated to 0.577350)
 		segment = 1
-		//then set x = (sqrt(3)*x-1)/(sqrt(3)+x)
-		sqrt3 := Sqrt(decimal.WithPrecision(calculatingPrecision), three)
-		sqrt3XMinus1 := decimal.WithPrecision(calculatingPrecision).Mul(sqrt3, x)
-		sqrt3XMinus1.Sub(sqrt3XMinus1, one)
-		sqrt3PlusX := x.Add(sqrt3, x)
-		x = sqrt3XMinus1.Quo(sqrt3XMinus1, sqrt3PlusX)
+
+		// then set x = (sqrt(3)*x-1)/(sqrt(3)+x)
+		sqrt3 := sqrt3(new(decimal.Big), ctx)
+		ctx.Mul(x0, sqrt3, x)    // sqrt(3) * x
+		ctx.Sub(x0, x0, one)     // sqrt(3)*x - 1
+		ctx.Add(sqrt3, sqrt3, x) // sqrt(3) + x
+		ctx.Quo(x0, x0, sqrt3)   // (sqrt(3)*x - 1) / (sqrt(3) + x)
+	} else {
+		x0.Copy(x)
 	}
 
-	//next we'll use argument halving technic
+	// next we'll use argument halving technic
 	// atan(y) = 2 atan(y/(1+sqrt(1+y^2)))
 	// we'll repeat this up to a point
 	// we have compeating operations at some
 	// point the sqrt causes a problem
-	// note (http://fredrikj.net/blog/2014/12/faster-high-precision-Atangents/)
+	// note (http:// fredrikj.net/blog/2014/12/faster-high-ctx-Atangents/)
 	// suggests that r= 8 times is a good value for
-	// precision with 1000s of digits to millions
+	// ctx with 1000s of digits to millions
 	// however it was easy to determine there is a
 	// better sliding window to use instead
 	// which is what we use as it turns out
-	// when the precision is large, a larger r value
+	// when the ctx is large, a larger r value
 	// compared to 8 is every effective
-	xFloat64, _ := x.Float64()
-	xFloat64 = stdMath.Abs(xFloat64)
-	//the formula simple but works a bit better then a fix value (8)
-	r := stdMath.Max(0, stdMath.Ceil(0.31554321636851*stdMath.Pow(float64(precision), 0.654095561044508)))
-	p := 1 / stdMath.Pow(2, r)
-	halfed := int(stdMath.Ceil(stdMath.Log(xFloat64/p) / stdMath.Log(2)))
+	xf, _ := x0.Float64()
+	xf = stdMath.Abs(xf)
+	// the formula simple but works a bit better then a fixed value (8)
+	r := stdMath.Max(0, stdMath.Ceil(0.31554321636851*stdMath.Pow(float64(ctx.Precision), 0.654095561044508)))
+	var p float64
+
+	// maxPrec is the largest precision value we can use bit shifts instead of
+	// math.Pow which is more expensive.
+	const maxPrec = 3286
+	if ctx.Precision <= maxPrec {
+		p = 1 / float64(uint64(1)<<uint64(r))
+	} else {
+		p = 1 / stdMath.Pow(2, r)
```

Comment:
This has me thinking, is stdMath.Pow(2,-r) better?

Reply:
Good idea.